torch_geometric/nn/conv/updated_gcn.py

```diff
@@ -0,0 +1,253 @@
+        self.explain = explain
+
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+        self.lin = Linear(in_channels, out_channels, bias=False,
+                          weight_initializer='glorot')
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(out_channels))
+        else:
+            self.register_parameter('bias', None)
+
+        if self.explain:
+            self.convert = Linear(in_channels, out_channels)
+            self.process_message = Sequential(LayerNorm(out_channels), ReLU())
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.explain:
+            uniform(self.in_channels, self.convert)
+        else:
+            self.lin.reset_parameters()
+            zeros(self.bias)
+            self._cached_edge_index = None
+            self._cached_adj_t = None
+
+    def forward(self, x: Tensor, edge_index: Adj,
+                edge_weight: OptTensor = None, message_scale=None,
+                message_replacement=None) -> Tensor:
+        """"""
+
+        if self.normalize:
+            if isinstance(edge_index, Tensor):
+                cache = self._cached_edge_index
+                if cache is None:
+                    edge_index, edge_weight = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_edge_index = (edge_index, edge_weight)
+                else:
+                    edge_index, edge_weight = cache[0], cache[1]
+
+            elif isinstance(edge_index, SparseTensor):
+                cache = self._cached_adj_t
+                if cache is None:
+                    edge_index = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_adj_t = edge_index
+                else:
+                    edge_index = cache
+
+        if not self.explain:
+            x = self.lin(x)
+
+        # propagate_type: (x: Tensor, edge_weight: OptTensor)
+        out = self.propagate(edge_index, x=x, edge_weight=edge_weight,
+                             message_scale=message_scale,
+                             message_replacement=message_replacement,
+                             size=None)
+
+        if self.bias is not None and not False and not self.explain:
+            out += self.bias
+
+        return out
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_index: Tensor,
+                edge_weight: OptTensor, message_scale,
+                message_replacement) -> Tensor:
+        if self.explain:
+            basis_messages = x_j if edge_weight is None else edge_weight.view(
+                -1, 1) * x_j
+            basis_messages = self.convert(basis_messages)
```

Comment:
It looks like it should be possible to out-source this functionality to its own function (between `message` and `aggregate`). We already do this to some extend here: https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py#L330-L341

I think one option that we have to make this GNN-agnostic is to leverage `MessagePassing.explain_message()` with the implementation of the current `MessagePassing` interface as default implementation, but which we can override for specific explainers. As such, the computation flow would be
```python
out = self.message(...)
out = self.explain_message(out)
out = self.aggregate(out, ...)
```
WDYT?

Reply:
I think that would be great if we follow your above mentioned notion of first letting the layer (GAT, GCN, RGCN, etc.) prepare the message then compute the `explain_message` by defining its own function in `MessagePassing` interface. This will not only save us from changing every single layer's implementation but also quite easy to implement. 

1. as we need to take into account relations for heterogeneous graphs, the implementation of computing `edge_weights` is somewhat different:-
For example, after computing messages, GCN explainable layer (updated_gcn.py) implements:

```
basis_messages = self.convert(basis_messages)
basis_messages = self.process_message(basis_messages)
```

and RGCN explainable layer (updated_rgcn.py) implements:

```
basis_messages = self.process_message(basis_messages)
```

So, is it possible if we pass in to this `explain_message()` function a flag which lets it to compute this explainable message for each type of graphs accordingly??

2. Since `self.convert()` and `self.process_message()` are instances of `torch_geometric.nn.dense.linear` and `torch.nn.Sequential` respectively, is it possible to use them in `MessagePassing` interface just by initializing them and set their initial weights somewhere in it, which let us employing these two functions (`self.convert()` and `self.process_message()`) in `explain_message()` I believe?? If yes, please suggest any way to achieve this or do you think any other best way to achieve this??

3. Do we have to register hooks for `self._explain_message_forward_pre_hooks.values()` and `self._explain_message_forward_hooks.values()` as `MessagePassing` interface implements for `message()` function given below:

```
for hook in self._message_forward_pre_hooks.values():
    res = hook(self, (msg_kwargs, ))
    if res is not None:
        msg_kwargs = res[0] if isinstance(res, tuple) else res
out = self.message(**msg_kwargs)
for hook in self._message_forward_hooks.values():
    res = hook(self, (msg_kwargs, ), out)
    if res is not None:
        out = res
```

If yes, can you help me with it just by writing here these two hooks statements for `explain_message()`, I'll take care of the rest?? That would be a great help, actually!!! :)

Reply:
1. `convert` is simply used to map messages to the desired output space, right? We could make this application optional inside `process_message`:
```python
if message.size(-1) != desired_size:
    message = self.lin(message)
```
where `self.lin = Linear(-1, desired_size)`.
2. I think we should take care of initializing these modules insice `GraphMask`:
```python
for conv in model.all_mp_modules:
    conv.message_lin = ...
    conv.explain_message = ...
```
3. I don't think we need any hooks for now.

Reply:
I think the goal here is to make this procedure GNN agnostic. As such, when we want to implement this inside a model's message function, we basically have to do this for every PyG GNN layer. Let me know if I might be misunderstanding something.

Reply:
@rusty1s Your above mentioned recommendations literally worked wonders :) Finally, devised the proposed `explain_message()` function. Have also accessed all the parameters for this function using `self.inspector.distribute('explain_message', coll_dict)` inside `MessagePassing` interface. Moreover, inside GraphMask, initialized `self.convert()` and `self.process_message()`, and accessed them inside `MessagePassing` interface while keeping the `self.convert()` optional to take care of mapping messages to the desired output space.

Furthermore, as discussed above, didn't implement any hooks for this `explain_message()` while employing the discussed computation flow inside `MessagePassing` interface:-

```
out = self.message(...)
out = self.explain_message(out, x_i, x_j, message_scale, message_replacement)
out = self.aggregate(out, ...)
```

Have generated explanations for predictions made by three different layers (GAT, GCN, RGCN) without even making any changes to their respective implementations. Used Datasets: Cora (Node Classification - GAT, GCN), Enzymes (Graph Classification - GAT, GCN), and AIFB (Node Classification - RGCN).

Reply:
Yeah, awesome! Can we add `explain_message` functionality in this PR (it's always a bit dangerous to make modifications to `MessagePassing` - we need to test that well), and then add support for `GraphMask` in a later PR?

Reply:
Sure. Have just added updated `MessagePassing` file (`updatedMessagePassing.py`) to this PR. One point to note here is that two parameters (`message_scale` and `message_replacement`) must be passed to both `forward()` and `propagate()` of each layer type (GAT, GCN, RGCN, etc.), as can be seen in files (`updated_gat.py`, `updated_gcn.py`, and `updated_rgcn.py`). If we don't do so, the following will surely be a result:- `__collect__()` of `MessagePassing` will fail to save these two parameters in `coll_dict`, which in turn, will fail to pass on these two parameters to `explain_message()` (after running `self.inspector.distribute('explain_message', coll_dict)`). Do let me know if you know some other way of passing these two parameters to `MessagePassing.propagate()` directly without touching the layer's existing implementation or else we must go with this approach of passing these two parameters to both `forward()` and `propagate()` of each layer type.

Furthermore, have taken `GraphMask.py` down because as discussed above, this PR is merely meant for adding `explain_message()` functionality to the `MessagePassing` interface. Besides, please take a deeper look at `self._gmask_explain` flag created inside `MessagePassing` interface, which helps triggering `explain_message` only when we work with GraphMask explainer or else will ensure the normal execution of the interface. This is to not disturb the original implementation of `MessagePassing` interface while not using this explanation technique.

All the changes made to the interface for this explanation technique to work can be seen inside the hash (#) blocks inside `MessagePassing` interface starts with (`# Support for GraphMask#######`) and ends with (`# block ended#######`). This is only to ease your review process for `updatedMessagePassing.py`.

Reply:
Resolving this conversation. Will continue in the main thread.